### PR TITLE
Add option to configure reduced precision math backend for SDPA

### PR DESCRIPTION
**Pull request by jianyuh**
Summary: Address https://github.com/pytorch/pytorch/issues/135778 by adding a global flag to configure whether using high precision or low precision for math backend of SDPA.

Test Plan: buck2 run mode/opt //scripts/feikou/llm:run_attn_kernels

Differential Revision: D62625515

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec @drisspg @mikaylagawarecki

### CODE REVIEW

File: torch/backends/cuda/__init__.py [hunk 1]
```diff
@@ -313,6 +313,24 @@ def math_sdp_enabled():
...
+def enable_math_sdp(enabled: bool):
+    r"""
+    .. warning:: This flag is beta and subject to change.
+
+    Enables or disables fp16/bf16 reduction in math scaled dot product attention.
+    """
+    torch._C.set_math_sdp_allow_fp16_bf16_reduction(enabled)
+
+
+def math_sdp_enabled():
```
Comment (drisspg) [inline torch/backends/cuda/__init__.py:325]: You are clashing with the existing functions, an you use a different name more specific to what this is doing

**Review by drisspg — COMMENTED** [review]
Left a comment I think we should test this as well

File: aten/src/ATen/Context.h [hunk 2]
```diff
@@ -390,6 +393,7 @@ class TORCH_API Context {
   bool enabled_mathSDP = true;
   bool enabled_cudnnSDP = true;
   bool enabled_overrideable = true;
+  bool allow_fp16_bf16_reduction_mathSDP = false;
```
Comment (drisspg) [inline aten/src/ATen/Context.h:396]: Cc @jbschlosser 
I think the default value should be switched. Since many people have voiced concerns over performance we should revert to the old behavior and make people turn manually set this to true

Comment (jbschlosser) [reply aten/src/ATen/Context.h:396]: Agree with Driss - imo let's revert to the old behavior as the default to fix the perf regression and make the high-precision behavior opt-in.

Comment (jbschlosser) [reply aten/src/ATen/Context.h:396]: After quite a bit of internal discussion, I'm adjusting my stance here. To summarize:
* We have precedent for providing high precision for composite ops by default and making low precision opt-in. The original PR can be considered a "bug fix" to bring consistency in following this precedent.
* I'm reading `allow_fp16_bf16_reduction_mathSDP` as "opt-in to low precision". So I think the default value should be `false`.

Unfortunately, this does mean those expecting the old behavior + performance have to explicitly change this flag.

**Comment by bghira** [comment]
i was going to thank you for backpedalling the fix but i'm not really sure the goal anymore here

File: aten/src/ATen/Context.h (hunk 2 above)
Comment (Beinsezii) [reply aten/src/ATen/Context.h:396]: >     I'm reading `allow_fp16_bf16_reduction_mathSDP` as "opt-in to low precision". So I think the default value should be `false`, in-line with the "bug fix" from the original PR.

So because the `allow_fp16_bf16_reduction_mathSDP` flag name implies "opt-in to low precision" it should be high precision by default?
The flag that was introduced in this same PR?
The flag we could rename right now before it lands in `release/2.5.0` and causes issue reports titled "*Help I updated the app and lost half my speed 6800 XT*" to open up on downstream repositories?

Comment (Beinsezii) [reply aten/src/ATen/Context.h:396]: Why not something easy to understand like `torch.set_sdpa_math_precision` keeping in spirit with `torch.set_float32_matmul_precision`? Then it could be "medium" or whatever you want by default.

Comment (jianyuh) [reply aten/src/ATen/Context.h:396]: @eqy Good point: we can have separate configurations on FP16/BF16 matmul reduced precision off/on with torch.backends.cuda.matmul.allow_fp16_reduced_precision_reduction = False/True. Currently the nested options are available for numerical testing.

For the original intention of https://github.com/pytorch/pytorch/pull/128922 , it's for improving the accuracy (and "Math SDPA reduced precision off" brings the best accuracy.

Comment (jianyuh) [reply aten/src/ATen/Context.h:396]: @Beinsezii  We can use other potential options (with better namings). Feel free to suggest better flag names.

> We have precedent for providing high precision for composite ops by default and making low precision opt-in.

From https://github.com/pytorch/pytorch/issues/76440 , https://github.com/pytorch/pytorch/pull/76509 , it matches what @jbschlosser  mentioned, we used  "opt-in to low precision" as a tradition in PyTorch (e.g., use TF32 for fp32 matmul, or use fp16/bf16 for math backend in SDPA), while the default one is the high precision (e.g., FP32 for fp32 matmul, and fp32 for math backend in SDPA).

**Comment by jianyuh** [comment]
@bghira We are providing an option to configure the low precision or high precision for math backend for SDPA. Originally we only by default use high precision for accuracy reasons (#128922). Feel free to suggest any better solutions here.

File: aten/src/ATen/Context.h (hunk 2 above)
Comment (Beinsezii) [reply aten/src/ATen/Context.h:396]: @jianyuh If my understanding is correct, the `scaled_dot_product_attention` function prefers the `FLASH_ATTENTION` and `EFFICIENT_ATTENTION` over the `MATH` kernels by default, despite the former two backends using bf16/fp16 like the old `MATH`. So if you're opting into pure `MATH` manually by disabling the fast kernels, to me it makes sense to also opt-into high precision ops. What having fp32 upcast MATH as the default effectively does is make the naive behavior of `scaled_dot_product_attention` use fast ops on NVIDIA and the MI300, but high precision ops on all the other cards which is an anti-pattern.

**Comment by bghira** [comment]
> @bghira We are providing an option to configure the low precision or high precision for math backend for SDPA. Originally we only by default use high precision for accuracy reasons (#128922). Feel free to suggest any better solutions here.

consistency is key and this isn't it @jianyuh 

SDPA isn't about having the best results, it is the fastest results. the same goes for efficient attn. it is not deterministic and the results don't even match vanilla attention.

it's mostly frustrating that it's just another nvidia-centric change without much regard for the impact on other platforms. furthermore we almost got the change to be opt-in instead of opt-out and then that was changed again without regard for impact on downstream users.

can you guys stop the NVIDIA-centric stuff over there at Meta? first xformers and now torch?

File: aten/src/ATen/Context.h (hunk 2 above)
Comment (jianyuh) [reply aten/src/ATen/Context.h:396]: @Beinsezii One issue for using fp16/bf16 for math backend (platform independent) is that it has much worse accuracy compared to flash attention or efficient attention backend.

Check https://github.com/pytorch/pytorch/pull/128922#issuecomment-2201176008

https://github.com/pytorch/pytorch/pull/128922#issuecomment-2251960439

Using high precision MATH makes the accuracy match more on flash attention or efficient attention with FP16/BF16.

In this PR, we intend to provide the *option* for the user to switch back to the less accurate but more efficient math backend, as you suggested.

**Comment by jianyuh** [comment]
@bghira  The original intention of #128922 is to trade off more on accuracy instead of performance/speed for math backend, regardless of HW platforms (NV/AMD GPU uses flash attention/efficient attention/flex attention not math backend). We want to provide an option for the user to configure this trade-off by the users in this PR.

**Comment by Beinsezii** [comment]
@jianyuh 
> (NV/AMD GPU uses flash attention/efficient attention/flex attention not math backend)

There's *two* AMD GPUs that can use flash/efficient kernels out of the 20 or so released in the last 5 years. The rest all use MATH. Also wouldn't MATH be the default for GPUs that aren't NVIDIA or AMD as well? Even some CPUs have f16 intrinsics now under AVX-512

You say "more accuracy instead of speed" but you're not talking +-15%, you're talking **half** of a card's throughput. I see this causing problems when downstream projects start depending on torch>=2.5.0 people have no idea why their GPUs are **half** speed because they weren't one of the 6 people reading this thread.

If the "internal discussions" have unequivocally dictated that it must be float32 by default, then I'd say at the very least add a nag to the console for a few versions. IIRC tf32 matmul change had a console nag for a while too.

**Comment by bghira** [comment]
> @bghira The original intention of #128922 is to trade off more on accuracy instead of performance/speed for math backend, regardless of HW platforms (NV/AMD GPU uses flash attention/efficient attention/flex attention not math backend). We want to provide an option for the user to configure this trade-off by the users in this PR.

ok, i see then that the issue is that nvidia's precision is too low by default and should be opt-in for low precision everywhere, even if the results aren't "as bad" as the MATH backend was. let's make things consistent one way or another for 2.5, and not have super different results across platforms.

**Comment by Beinsezii** [comment]
If we always want the highest precision as the default values then
```
torch.backends.cuda.matmul.allow_fp16_reduced_precision_reduction
torch.backends.cuda.matmul.allow_bf16_reduced_precision_reduction
torch.backends.cuda.matmul.allow_tf32
```
Should all default to `False`.

I'm also surprised SDPA needs its own strange flag instead of just using the already existing fp16/bf16 reduction flags?

**Comment by jbschlosser** [comment]
> If we always want the highest precision as the default values then
> torch.backends.cuda.matmul.allow_fp16_reduced_precision_reduction
> torch.backends.cuda.matmul.allow_bf16_reduced_precision_reduction
> torch.backends.cuda.matmul.allow_tf32
> Should all default to False.

I'd agree that they should all ideally be set to False to be consistent with the "high precision by default" policy. Note that this is the case for `allow_tf32`, and ideally we would bring the other flags in-line with this eventually. There is an effort in #125888 to cleanup / refine precision-related flags, and perhaps a follow-up will address the inconsistency here.

> If the "internal discussions" have unequivocally dictated that it must be float32 by default, then I'd say at the very least add a nag to the console for a few versions. IIRC tf32 matmul change had a console nag for a while too.

Note that the "console nag for a release" approach would result in quite a bit of flip-flopping for this behavior. Imo it's preferable to get to the desired end state (high precision by default) sooner rather than later to reduce overall confusion. Arguably, a slower rollout for the change (possibly including a console nag for a release) would have been a better way to make this transition in the first place, but now that the behavior has been changed (albeit in a sudden BC-breaking way), I think it's better to keep the change while making it opt-out.

> I'm also surprised SDPA needs its own strange flag instead of just using the already existing fp16/bf16 reduction flags?

The idea is that SDPA should be considered as an op on its own, with its own precision properties / guarantees, and "greater than the sum of its parts". To that end, an SDPA-specific flag makes sense over reusing the matmul flags.

**Review by jbschlosser — COMMENTED** [review]
still need tests :) and imo it would be valuable to update `docs/source/notes/numerical_accuracy.rst` with both a description of the new flag and how its behavior intersects with the existing flags

**Comment by bghira** [comment]
it wasn't broken in a release. just on main. so the change isn't something to commit to yet. it was a mistake to be fixed

**Comment by Beinsezii** [comment]
> Note that the "console nag for a release" approach would result in quite a bit of flip-flopping for this behavior. Imo it's preferable to get to the desired end state (high precision by default) sooner rather than later to reduce overall confusion. Arguably, a slower rollout for the change (possibly including a console nag for a release) would have been a better way to make this transition in the first place, but now that the behavior has been changed (albeit in a sudden BC-breaking way), I think it's better to keep the change while making it opt-out.

We're thinking two different things I believe. I was saying more if it's going to be high precision by default, have a console nag for a few versions that says something like
```
Using float32 for SDPBackend.MATH. To change this behavior set
torch.nn.attention.allow_bf16_fp16_reduction_mathsdp = True
```
that shows the new default and how to get the old behavior. I thought something similar was implemented for the tf32 matmul change?

File: test/test_transformers.py [hunk 3]
```diff
@@ -849,6 +849,42 @@ def test_encoder_is_causal(self):
 
         self.assertEqual(masked_output, is_causal_output)
 
+    @onlyCUDA
+    def test_math_backend_high_precision(self):
+        xq = torch.rand([1, 6432, 2, 80], device="cuda", dtype=torch.bfloat16) * 5
+        xk = torch.rand([1, 6432, 2, 80], device="cuda", dtype=torch.bfloat16) * 5
+        xv = torch.randn([1, 6432, 2, 80], device="cuda", dtype=torch.bfloat16)
```
Comment (jbschlosser) [inline test/test_transformers.py:856]: can you get away with smaller sizes during testing?

File: test/test_transformers.py [hunk 4]
```diff
@@ -849,6 +849,42 @@ def test_encoder_is_causal(self):
 
         self.assertEqual(masked_output, is_causal_output)
 
+    @onlyCUDA
+    def test_math_backend_high_precision(self):
```
Comment (jbschlosser) [inline test/test_transformers.py:853]: the test coverage I'm looking for it just verifying that setting the flag does indeed turn on low precision mode. This might be as easy as comparing the differences between:
1. flag off SDPA math vs. the "gold standard"
2. flag on SDPA math vs. the "gold standard"

the difference for 1 should be substantially < 2 if the flag is working correctly

I see you're using flash as the gold standard, but is it better to use float64 math? If flash is preferred, we'll need to decorate this test so it only runs on platforms where flash is supported e.g.
https://github.com/pytorch/pytorch/blob/599735415124d78f2b1766681da1547a62681c8d/test/test_transformers.py#L1393-L1395

File: test/test_transformers.py [hunk 5]
```diff
@@ -849,6 +849,42 @@ def test_encoder_is_causal(self):
...
+        torch.backends.cuda.allow_fp16_bf16_reduction_math_sdp(True)
+        sdp_math_low_prec_out = scaled_dot_product_attention(xq, xk, xv, mask, SDPBackend.MATH)
+        torch.backends.cuda.allow_fp16_bf16_reduction_math_sdp(False)
+        sdp_math_high_prec_out = scaled_dot_product_attention(xq, xk, xv, mask, SDPBackend.MATH)
+        sdp_flash_out = scaled_dot_product_attention(xq, xk, xv, mask, SDPBackend.FLASH_ATTENTION)
+
+        torch.testing.assert_close(sdp_math_high_prec_out, sdp_flash_out, atol=1e-2, rtol=1e-2)
+
+        try:
+            torch.testing.assert_close(sdp_math_low_prec_out, sdp_flash_out, atol=1e-2, rtol=1e-2)
```
Comment (jbschlosser) [inline test/test_transformers.py:882]: as is, both of these assertions could pass and we wouldn't know whether the flag controls precision or not. see my prior suggestion for what I'm looking for from this test

File: docs/source/notes/numerical_accuracy.rst [hunk 6]
```diff
@@ -110,6 +110,13 @@ reduced-precision reductions are problematic, they can be turned off with
 
 For more information see :ref:`allow_fp16_reduced_precision_reduction<fp16reducedprecision>` and :ref:`allow_bf16_reduced_precision_reduction<bf16reducedprecision>`
 
+Reduced Precision Reduction for FP16 and BF16 in Scaled Dot Product Attention (SDPA)
+------------------------------------------------------------------------------------
+The SDPA math backend, when using FP16/BF16 inputs, exhibits significant numerical errors due to the low-precision intermediate buffers. To mitigate this issue, the default behavior now involves converting FP16/BF16 inputs to FP32. Computations are performed in FP32/TF32, and the final FP32 results are then converted back to FP16/BF16. This will improved numerical accuracy for math backend with FP16/BF16 inputs, but potentially caused the performance regression in the math backend as computations shift from FP16/BF16 BMM to FP32/TF32 BMM/Matmul.
```
Comment (jbschlosser) [inline docs/source/notes/numerical_accuracy.rst:115]: some doc nits:
```suggestion
A naive SDPA math backend, when using FP16/BF16 inputs, can accumulate significant numerical errors due to the usage of low-precision intermediate buffers. To mitigate this issue, the default behavior now involves upcasting FP16/BF16 inputs to FP32. Computations are performed in FP32/TF32, and the final FP32 results are then downcasted back to FP16/BF16. This will improve numerical accuracy of the final output for the math backend with FP16/BF16 inputs, but increases memory usage and may cause performance regressions in the math backend as computations shift from FP16/BF16 BMM to FP32/TF32 BMM/Matmul.
```

**Review by jbschlosser — COMMENTED** [review]
Thanks for adding the test! my only concern is that it's not clear from the test whether the flag actually controls the precision of the SDPA math backend, which is what I'm looking to verify with the test coverage. See my suggestion below for how to address this.

otherwise, I'm good with it. We need to get it cherry-picked into the RC branch ASAP

**Comment by jbschlosser** [comment]
> We're thinking two different things I believe. I was saying more if it's going to be high precision by default, have a console nag for a few versions that says something like

Ah you're right - I was considering a pre-change console nag and it sounds like you're proposing a console nag signaling that the behavior has already changed. Note that our policy now is to avoid non-actionable warnings when possible. There will be a BC-breaking note in the release notes for the default behavior change.

File: test/test_transformers.py (hunk 5 above)
Comment (jianyuh) [reply test/test_transformers.py:882]: @jbschlosser for the 2nd assert on sdp_math_low_prec_out, we actually want to assert failure with try / except.

File: test/test_transformers.py (hunk 3 above)
Comment (jianyuh) [reply test/test_transformers.py:856]: Changed 6432 to 128 and we can capture the issue with the test.

File: test/test_transformers.py (hunk 4 above)
Comment (jianyuh) [reply test/test_transformers.py:853]: Update the flash backend with float64 math as you suggested.

Comment (jianyuh) [reply test/test_transformers.py:853]: We used `sdp_math_low_prec_out` for 1, and `sdp_math_high_prec_out` for 2.

The test shows that 2 can pass the 1st torch.testing.assert_close; but 1 will fail.

**Comment by jianyuh** [comment]
Hi @jbschlosser 

> my only concern is that it's not clear from the test whether the flag actually controls the precision of the SDPA math backend, which is what I'm looking to verify with the test coverage.

You can check the test, where we turn on/off the flag and compare the result against the golden standard.

```
        torch.backends.cuda.allow_fp16_bf16_reduction_math_sdp(True)
        sdp_math_low_prec_out = scaled_dot_product_attention(xq, xk, xv, mask, SDPBackend.MATH)
        torch.backends.cuda.allow_fp16_bf16_reduction_math_sdp(False)
        sdp_math_high_prec_out = scaled_dot_product_attention(xq, xk, xv, mask, SDPBackend.MATH)
```

File: test/test_transformers.py (hunk 4 above)
Comment (jbschlosser) [reply test/test_transformers.py:856]: okay looks good, thanks for clarifying! FYI there's some sugar you can use when you expect an error:
```python
with self.assertRaisesRegex(AssertionError, "Tensor-likes are not close"):
    ...
```

**Review by jbschlosser — APPROVED** [review]
